sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala

```diff
@@ -207,15 +210,31 @@ class ParquetFileFormat
 
       val sharedConf = broadcastedHadoopConf.value.value
 
-      val fileFooter = if (enableVectorizedReader) {
-        // When there are vectorized reads, we can avoid reading the footer twice by reading
-        // all row groups in advance and filter row groups according to filters that require
-        // push down (no need to read the footer metadata again).
-        ParquetFooterReader.readFooter(sharedConf, file, ParquetFooterReader.WITH_ROW_GROUPS)
+      // When there are vectorized reads, we can avoid
+      // 1. opening the file twice by transfering the SeekableInputStream
+      // 2. reading the footer twice by reading all row groups in advance and filter row groups
+      //    according to filters that require push down
+      val metadataFilter = if (enableVectorizedReader) {
+        HadoopReadOptions.builder(sharedConf, filePath)
+          .withRange(file.start, file.start + file.length)
+          .build.getMetadataFilter
       } else {
-        ParquetFooterReader.readFooter(sharedConf, file, ParquetFooterReader.SKIP_ROW_GROUPS)
+        ParquetMetadataConverter.SKIP_ROW_GROUPS
       }
 
+      val readOptions = HadoopReadOptions.builder(sharedConf, filePath)
+        .withMetadataFilter(metadataFilter).build
+
+      val inputFile = HadoopInputFile.fromStatus(file.fileStatus, sharedConf)
+      val inputStream = inputFile.newStream()
+      val fileReader = ParquetFileReader.open(inputFile, readOptions, inputStream)
+      val fileFooter = fileReader.getFooter
+      if (enableVectorizedReader) {
+        // Keep the file input stream open so it can be reused later
+        fileReader.detachFileInputStream()
+      }
+      fileReader.close()
```

Comment:
`fileReader.close()` should in finally block

Reply:
rewrite by using `Utils.tryWithResource`